src/test/httpserver_tests.cpp

```diff
@@ -13,26 +15,34 @@ BOOST_AUTO_TEST_CASE(test_query_parameters)
 {
     std::string uri {};
 
+    evhttp_request* evreq = evhttp_request_new(nullptr, nullptr);
```

Comment:
See https://github.com/bitcoin/bitcoin/pull/24012 - I couldn't get mocking an `evhttp_request*` past the ASan test. That's why in #24098 I introduced the `GetQueryParameterFromUri()` helper function: it could easily be unit tested, and then `HTTPRequest::GetQueryParameter()` was just a simple wrapper left untested.

Reply:
If I bring it back I won't be able to access to m_uri_parsed from the helper function, unless there's another alternative, I'll have to roll-back to the original approach of the PR.

Reply:
What do you think about this approach? I think it's quite minimal:

<details>
<summary>git diff</summary>

```diff
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 942caa042..6a6176a78 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -15,6 +15,7 @@
 #include <rpc/protocol.h> // For HTTP status codes
 #include <shutdown.h>
 #include <sync.h>
+#include <util/check.h>
 #include <util/strencodings.h>
 #include <util/syscall_sandbox.h>
 #include <util/system.h>
@@ -253,6 +254,14 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         return;
     }
 
+    // Early reject invalid URI
+    if (!hreq->GetURIParsed()) {
+        LogPrint(BCLog::HTTP, "HTTP request from %s rejected: Invalid URI\n",
+                 hreq->GetPeer().ToStringAddrPort());
+        hreq->WriteReply(HTTP_BAD_REQUEST, "Invalid URI");
+        return;
+    }
+
     LogPrint(BCLog::HTTP, "Received a %s request for %s from %s\n",
              RequestMethodString(hreq->GetRequestMethod()), SanitizeString(hreq->GetURI(), SAFE_CHARS_URI).substr(0, 100), hreq->GetPeer().ToStringAddrPort());
 
@@ -535,18 +544,22 @@ void HTTPEvent::trigger(struct timeval* tv)
     else
         evtimer_add(ev, tv); // trigger after timeval passed
 }
-HTTPRequest::HTTPRequest(struct evhttp_request* _req, bool _replySent) : req(_req), replySent(_replySent)
+HTTPRequest::HTTPRequest(struct evhttp_request* _req, bool _replySent) :
+        req{_req},
+        m_uri_parsed{evhttp_uri_parse(evhttp_request_get_uri(req))},
+        replySent{_replySent}
 {
 }
 
 HTTPRequest::~HTTPRequest()
 {
+    // evhttpd cleans up the request, as long as a reply was sent.
     if (!replySent) {
         // Keep track of whether reply was sent to avoid request leaks
         LogPrintf("%s: Unhandled request\n", __func__);
         WriteReply(HTTP_INTERNAL_SERVER_ERROR, "Unhandled request");
     }
-    // evhttpd cleans up the request, as long as a reply was sent.
+    if(m_uri_parsed) evhttp_uri_free(m_uri_parsed);
 }
 
 std::pair<bool, std::string> HTTPRequest::GetHeader(const std::string& hdr) const
@@ -665,14 +678,13 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
 
 std::optional<std::string> HTTPRequest::GetQueryParameter(const std::string& key) const
 {
-    const char* uri{evhttp_request_get_uri(req)};
-
-    return GetQueryParameterFromUri(uri, key);
+    if (!Assume(m_uri_parsed)) return std::nullopt;  // http_request_cb should have checked that m_uri_parsed is not nullptr
+    return GetQueryParameterFromUri(m_uri_parsed, key);
 }
 
-std::optional<std::string> GetQueryParameterFromUri(const char* uri, const std::string& key)
+std::optional<std::string> GetQueryParameterFromUri(const evhttp_uri* uri_parsed, const std::string& key)
 {
-    evhttp_uri* uri_parsed{evhttp_uri_parse(uri)};
+    assert(uri_parsed);
     const char* query{evhttp_uri_get_query(uri_parsed)};
     std::optional<std::string> result;
 
@@ -689,7 +701,6 @@ std::optional<std::string> GetQueryParameterFromUri(const char* uri, const std::
         }
         evhttp_clear_headers(&params_q);
     }
-    evhttp_uri_free(uri_parsed);
 
     return result;
 }
diff --git a/src/httpserver.h b/src/httpserver.h
index 036a39a02..84906e121 100644
--- a/src/httpserver.h
+++ b/src/httpserver.h
@@ -14,6 +14,7 @@ static const int DEFAULT_HTTP_WORKQUEUE=16;
 static const int DEFAULT_HTTP_SERVER_TIMEOUT=30;
 
 struct evhttp_request;
+struct evhttp_uri;
 struct event_base;
 class CService;
 class HTTPRequest;
@@ -57,6 +58,7 @@ class HTTPRequest
 {
 private:
     struct evhttp_request* req;
+    evhttp_uri* m_uri_parsed;
     bool replySent;
 
 public:
@@ -75,6 +77,9 @@ public:
      */
     std::string GetURI() const;
 
+    /** Get libevent parsed URI, which is nullptr if uri parsing failed */
+    const evhttp_uri* GetURIParsed() const { return m_uri_parsed; }
+
     /** Get CService (address:ip) for the origin of the http request.
      */
     CService GetPeer() const;
@@ -135,10 +140,10 @@ public:
  *
  * Helper function for HTTPRequest::GetQueryParameter.
  *
- * @param[in] uri is the entire request uri
+ * @param[in] uri_parsed is the request uri after parsing it with evhttp_uri_parse
  * @param[in] key represents the query parameter of which the value is returned
  */
-std::optional<std::string> GetQueryParameterFromUri(const char* uri, const std::string& key);
+std::optional<std::string> GetQueryParameterFromUri(const evhttp_uri* uri_parsed, const std::string& key);
 
 /** Event handler closure.
  */

```
</details>

Reply:
I see, ok, it makes sense, let me make some changes.